source/extensions/filters/http/ratelimit/ratelimit.cc

```diff
@@ -161,6 +164,18 @@ void Filter::complete(Filters::Common::RateLimit::LimitStatus status,
     break;
   }
 
+  if (config_->enableResponseHeaders()) {
+    Http::ResponseHeaderMapPtr rate_limit_headers =
+        RateLimitHeaders::create(std::move(descriptor_statuses));
+    if (response_headers_to_add_ == nullptr) {
+      response_headers_to_add_ = Http::ResponseHeaderMapImpl::create();
+    }
+    Http::HeaderUtility::addHeaders(*response_headers_to_add_, *rate_limit_headers);
+    rate_limit_headers = nullptr;
```

Comment:
interesting. Why are we resetting the rate_limit_headers here and the descriptor_statuses below?

Reply:
RateLimitHeaders::create(std::move(descriptor_statuses)); on line 168 moved the ownership of the descriptor_statuses pointer to RateLimitHeaders, which is consuming the pointer. if this feature is disabled, we still need to consume the pointer, thus L176. As for rate_limit_headers - we are the owners of it, we have to consume it after using